File: Commons/Math/src/main/java/com/greentree/commons/math/vector/AbstractFloatVector.java
package com.greentree.commons.math.vector;

import java.util.Iterator;

import com.greentree.commons.math.Mathf;

public abstract class AbstractFloatVector extends AbstractVector implements Iterable<Float> {


	private static final long serialVersionUID = 1L;

	public AbstractFloatVector add(AbstractFloatVector v) {
		return add(v, this);
	}

	public <T extends AbstractFloatVector> T add(AbstractFloatVector v, T dest) {
		checkSize(dest);
		checkSize(v);
		for(int i = 0; i < v.size(); i++)
			dest.set(i, get(i) + v.get(i));
		return dest;
	}
	public float distance(AbstractFloatVector v) {
		return Mathf.sqrt(distanceSqr(v));
	}

	public float distanceSqr(AbstractFloatVector v) {
		checkSize(v);
		final var s = new VectorNf(size());
		sub(v, s);
		return s.lengthSquared();
	}

	public float dot(AbstractFloatVector v) {
		checkSize(v);
		var res = 0f;
		for(int i = 0; i < size(); i++)
			res += v.get(i) * get(i);
		return res;
	}

	@Override
	public boolean equals(Object obj) {
		if(this == obj) return true;
		if(obj == null || getClass() != obj.getClass()) return false;
		var other = (AbstractVectorNf) obj;
		if(size() != other.size()) return false;
		for(int i = 0; i < size(); i++)
			if(Mathf.equals(get(i), other.get(i)))
				return false;
		return true;
	}
	public abstract float get(int index);

	@Override
	public int hashCode() {
		var res = 1f;
		for(int i = 0; i < size(); i++)
			res = 31 * res + get(i);
		return Float.floatToIntBits(res);
	}

	@Override
	public Iterator<Float> iterator() {
		return new AbstractFloatVectorIterator();
	}

	@Override
	public float lengthSquared() {
		var lengthSquared = 0.0f;
		for(int i = 0; i < size(); i++) {
			final var x = get(i);
			lengthSquared += x * x;
		}
		return lengthSquared;
	}

	public AbstractFloatVector mul(AbstractFloatVector v) {
		return mul(v, this);
	}

	public <T extends AbstractFloatVector> T mul(AbstractFloatVector v, T dest) {
		checkSize(dest);
		checkSize(v);
		for(int i = 0; i < v.size(); i++)
			dest.set(i, get(i) * v.get(i));
		return dest;
	}

	public AbstractFloatVector mul(float k) {
		return mul(k, this);
	}

	public <T extends AbstractFloatVector> T mul(float k, T dest) {
		checkSize(dest);
		for(int i = 0; i < size(); i++)
			dest.set(i, k * get(i));
		return dest;
	}

	public AbstractFloatVector normalize() {
		return normalize(1, this);
	}

	public AbstractFloatVector normalize(float radius) {
		return normalize(radius, this);
	}
	public <T extends AbstractFloatVector> T normalize(float radius, T dest) {
		var len = length();
		if(len > 1E-9)
			return mul(radius / len, dest);
		return dest;
	}

	public <T extends AbstractFloatVector> T normalize(T dest) {
		return normalize(1, dest);
	}

	public AbstractFloatVector projection(AbstractFloatVector norm) {
		return projection(norm, this);
	}

	public <T extends AbstractFloatVector>  T projection(AbstractFloatVector norm, T dest) {
		checkSize(norm);
		checkSize(dest);
		return norm.normalize(dot(norm), dest);
	}

	public abstract void set(int index, float x);

	public AbstractFloatVector sub(AbstractFloatVector v) {
		return sub(v, this);
	}

	public <T extends AbstractFloatVector> T sub(AbstractFloatVector v, T dest) {
		checkSize(dest);
		checkSize(v);
		for(int i = 0; i < v.size(); i++)
			dest.set(i, get(i) - v.get(i));
		return dest;
	}

	public final class AbstractFloatVectorIterator implements Iterator<Float> {

		private int index = 0;

		@Override
		public boolean hasNext() {
			return index < size();
		}

		@Override
		public Float next() {
			return get(index++);
		}

	}

}

File: Commons/Math/src/main/kotlin/com/greentree/commons/math/vector/AbstractFloatVector.kt
package com.greentree.commons.math.vector

import com.greentree.commons.math.Mathf
import com.greentree.commons.math.vector.AbstractVector.Companion.checkSize

interface AbstractFloatVector : AbstractVector<Float> {

	override fun dot(other: AbstractVector<out Float>): Float {
		checkSize(this, other)
		return zip(other).map { (x, y) -> x * y }.sum()
	}

	fun projection(normal: AbstractFloatVector): AbstractFloatVector {
		return normal.normalize(dot(normal))
	}

	override fun magnitude(length: Number): AbstractFloatVector {
		if (lengthSquared() < Mathf.EPS)
			return this
		return normalize(length)
	}

	override fun normalize(length: Number): AbstractFloatVector {
		return this * length / length()
	}

	operator fun plus(other: Number): AbstractFloatVector {
		return plus(other.toFloat())
	}

	operator fun div(other: Number): AbstractFloatVector {
		return div(other.toFloat())
	}

	operator fun times(other: Number): AbstractFloatVector {
		return times(other.toFloat())
	}

	operator fun minus(other: Number): AbstractFloatVector {
		return minus(other.toFloat())
	}

	override fun plus(other: Float): AbstractFloatVector {
		return plus(newVector(other))
	}

	override fun div(other: Float): AbstractFloatVector {
		return div(newVector(other))
	}

	override fun times(other: Float): AbstractFloatVector {
		return times(newVector(other))
	}

	override fun minus(other: Float): AbstractFloatVector {
		return minus(newVector(other))
	}

	override fun plus(other: AbstractVector<out Float>): AbstractFloatVector {
		checkSize(this, other)
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = get(i) + other[i]
		return dest
	}

	override fun div(other: AbstractVector<out Float>): AbstractFloatVector {
		checkSize(this, other)
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = get(i) / other[i]
		return dest
	}

	override fun times(other: AbstractVector<out Float>): AbstractFloatVector {
		checkSize(this, other)
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = get(i) * other[i]
		return dest
	}

	override fun minus(other: AbstractVector<out Float>): AbstractFloatVector {
		checkSize(this, other)
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = get(i) - other[i]
		return dest
	}

	override fun unaryMinus(): AbstractFloatVector {
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = -get(i)
		return dest
	}

	override fun unaryPlus(): AbstractFloatVector {
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = get(i)
		return dest
	}

	override fun lengthSquared(): Float {
		return this.map { it * it }.sum()
	}

	override fun toMutable(): AbstractMutableFloatVector {
		val dest = VectorNf(size)
		for (i in 0 until size) dest[i] = get(i)
		return dest
	}
}

operator fun Number.plus(other: AbstractVector<out Float>): AbstractFloatVector {
	return newVector(this, other.size) + other
}

operator fun Number.times(other: AbstractVector<out Float>): AbstractFloatVector {
	return newVector(this, other.size) * other
}

operator fun Number.minus(other: AbstractVector<out Float>): AbstractFloatVector {
	return newVector(this, other.size) - other
}

operator fun Number.div(other: AbstractVector<out Float>): AbstractFloatVector {
	return newVector(this, other.size) / other
}

fun newVector(value: Number, size: Int): AbstractFloatVector {
	return newVector(value.toFloat(), size)
}

fun AbstractFloatVector.newVector(value: Float): AbstractFloatVector {
	return OneComponentFloatVector(value, size)
}

fun AbstractFloatVector.newVector(value: Number): AbstractFloatVector {
	return newVector(value.toFloat(), size)
}

fun newVector(value: Float, size: Int): AbstractFloatVector {
	return OneComponentFloatVector(value, size)
}

class OneComponentFloatVector(val value: Float, override val size: Int) : AbstractFloatVector {

	override fun get(index: Int): Float = value
}